#' Load all data required for NanoSeq analysis
#' 
#' Package requirements (load before use): tidyverse, vcfR, BSgenome package corresponding to the reference genome used in the NanoSeq analysis
#'
#' This function loads all the files for analysis of NanoSeq data. For more information regarding NanoSeq output files, refer to the "output" section in the [NanoSeq GitHub](https://github.com/cancerit/NanoSeq).
#'
#' @param dirs A character vector of the directories containing NanoSeq results to load.
#' @param sample_names A character vector of the sample names to assign to the results, in the same order as the directories in 'dirs'.
#' @param BSgenomepackagename A string of a BSgenome package corresponding to the reference genome used in the Nanoseq analysis (e.g., "BSgenome.Hsapiens.UCSC.hg38" or "BSgenome.Mmusculus.UCSC.mm10"). This is used to calculate the genome trinucleotide background and corrected substitution counts and burdens.
#' @param BSgenomecontigs A vector of numeric indices of the contigs of the BSgenome package from which to calculate the genome trinucleotide background (e.g., 1:24 for BSgenome.Hsapiens.UCSC.hg38, or 1:21 for BSgenome.Mmusculus.UCSC.mm10).
#' @return A list of data files with all the relevant results generated by NanoSeq.
#' * sample_id: A vector of all sample IDs that were loaded
#' * vcf_snp.fix: List (one object per sample) containing the fixed information (fix) from the SNP vcf (only FILTER = PASS mutations)
#' * vcf_indel.fix: List (one object per sample) containing the fixed information (fix) from the indel vcf (only FILTER = PASS mutations)
#' * vcf_indel.gt: List (one object per sample) containing the genotype information (gt) from the indel vcf (only FILTER = PASS mutations)
#' * trinuc_bg_ratio: Data frame of the sample trinucleotide background (i.e. number of interrogated bases for each trinucleotide context), the genome trinucleotide background (i.e. number of each trinucleotide context), and the normalized ratio of these. Columns: sample, tri (trinucleotide context), sample_tri_bg, genome_tri_bg, ratio2genome.
#' * trinuc_bg.sigfit: Data frame in sigfit format of the trinucleotide background (i.e. number of interrogated bases for each trinucleotide context), with one row per sample and one column per trinucleotide context.
#' * observed_corrected_trinuc_counts: Data frame of observed and corrected mutation counts (for all mutations and for unique mutations). Columns: sample, tri (trinucleotide context), trint_subst_observed, trint_subst_unique_observed, ratio2genome, trint_subst_corrected, trint_subst_unique_corrected.
#' * observed_trinuc_counts.sigfit: Data frame in sigfit format of unique observed mutation counts, with one row per sample and one column per trinucleotide substitution context.
#' * mutation_burden: The total number of observed and corrected mutations, total number of observed and corrected interrogated bases (note: observed and corrected are the same), observed and corrected mutation burdens, observed and corrected lower and upper confidence intervals of mutation counts, and observed and corrected lower and upper confidence intervals of mutation burdens, with one row per sample. All these statistics include all mutations, not just unique mutations.
#' * purine_trinuc_mismatches: Data frame of the number of single-strand consensus purine mismatches. Columns: sample, tri (trinucleotide context), value.
#' * pyrimidine_trinuc_mismatches: Data frame of the number of single-strand consensus pyrimidine mismatches. Columns: sample, tri (trinucleotide context), value.
#' * estimated_error_rates: Data frame of the probability of having independent errors affecting both strands and resulting in a false-positive double-strand mutation and the number of estimated false positive double-strand mutations, based on the independent error rates in the purine channels. Columns: sample, total_error_rate, total_errors.
#' @export

load_nanoseq_data <- function(dirs, sample_names, BSgenomepackagename, BSgenomecontigs) {
  
  #Load BSgenome package
  library(BSgenomepackagename,character.only=TRUE)
  
  # Initialize dataframes for results
  vcf_snp.fix <- list()
  vcf_indel.fix <- list()
  vcf_indel.gt <- list()
  results.trint_counts_and_ratio2genome <- list()
  results.trint_subs_obs_corrected <- list()
  results.mut_burden <- list()
  results.SSC_mismatches_purine <- list()
  results.SSC_mismatches_pyrimidine <- list()
  results.estimated_error_rates <- list()

  # Order of trinucleotide context labels
  trint_subs_labels <- c("ACA>AAA","ACC>AAC","ACG>AAG","ACT>AAT","CCA>CAA","CCC>CAC","CCG>CAG","CCT>CAT","GCA>GAA","GCC>GAC","GCG>GAG","GCT>GAT","TCA>TAA","TCC>TAC","TCG>TAG","TCT>TAT","ACA>AGA","ACC>AGC","ACG>AGG","ACT>AGT","CCA>CGA","CCC>CGC","CCG>CGG","CCT>CGT","GCA>GGA","GCC>GGC","GCG>GGG","GCT>GGT","TCA>TGA","TCC>TGC","TCG>TGG","TCT>TGT","ACA>ATA","ACC>ATC","ACG>ATG","ACT>ATT","CCA>CTA","CCC>CTC","CCG>CTG","CCT>CTT","GCA>GTA","GCC>GTC","GCG>GTG","GCT>GTT","TCA>TTA","TCC>TTC","TCG>TTG","TCT>TTT","ATA>AAA","ATC>AAC","ATG>AAG","ATT>AAT","CTA>CAA","CTC>CAC","CTG>CAG","CTT>CAT","GTA>GAA","GTC>GAC","GTG>GAG","GTT>GAT","TTA>TAA","TTC>TAC","TTG>TAG","TTT>TAT","ATA>ACA","ATC>ACC","ATG>ACG","ATT>ACT","CTA>CCA","CTC>CCC","CTG>CCG","CTT>CCT","GTA>GCA","GTC>GCC","GTG>GCG","GTT>GCT","TTA>TCA","TTC>TCC","TTG>TCG","TTT>TCT","ATA>AGA","ATC>AGC","ATG>AGG","ATT>AGT","CTA>CGA","CTC>CGC","CTG>CGG","CTT>CGT","GTA>GGA","GTC>GGC","GTG>GGG","GTT>GGT","TTA>TGA","TTC>TGC","TTG>TGG","TTT>TGT")

  genome_freqs_labels <- str_sub(trint_subs_labels,1,3)
  
  #Helper function to reduce 64 to 32 trinucleotide frequency with central pyrimidine. Input is integer array with named elements that results from the trinucleotideFrequency function of Biostrings.
  trinucleotide64to32 <- function(x){
    suppressPackageStartupMessages(library(Biostrings))
    trinucleotides_64 <- apply(expand.grid(c("A","C","G","T"),c("A","C","G","T"),c("A","C","G","T")),1,paste,collapse="")
    trinucleotides_32_pyr <- apply(expand.grid(c("A","C","G","T"),c("C","T"),c("A","C","G","T")),1,paste,collapse="")
    trinucleotides_32_pur <- setdiff(trinucleotides_64,trinucleotides_32_pyr)
    
    y <- x[trinucleotides_32_pur]
    x <- x[trinucleotides_32_pyr]
    names(y) <- reverseComplement(DNAStringSet(names(y)))
    result <- merge(x,y,by=0)
    row.names(result) <- result[,1]
    return(apply(result[,-1],1,sum))
  }
  
  #Calculate reference genome trinucloetide counts, reduced to 32 trinucleotide contexts
  genome_trinuc_counts <- NULL
  for(i in BSgenomecontigs){
    genome_trinuc_counts <- rbind(genome_trinuc_counts,trinucleotideFrequency(eval(parse(text=BSgenomepackagename))[[i]]))
  }
  genome_trinuc_counts <- genome_trinuc_counts %>% colSums %>% trinucleotide64to32 %>% as.data.frame %>% rownames_to_column("tri")
  colnames(genome_trinuc_counts)[2] <- "genome_tri_bg"
  
  message("Loading sample data...")
  for (i in 1:length(dirs)) {
    
    dir <- dirs[i]
    sample_name <- sample_names[i]

    # Load vcf files
    vcf_snp <- read.vcfR(paste0(dir,"/results.muts.vcf.gz"),verbose=FALSE)
    vcf_snp.fix[[sample_name]] <- data.frame(vcf_snp@fix) %>% filter(FILTER == "PASS")
    
    vcf_indel <- read.vcfR(paste0(dir,"/results.indel.vcf.gz"),verbose=FALSE)
    vcf_indel.fix[[sample_name]] <- data.frame(vcf_indel@fix) %>% filter(FILTER == "PASS")
    vcf_indel.gt[[sample_name]] <- data.frame(vcf_indel@gt) %>% filter(data.frame(vcf_indel@fix)$FILTER == "PASS")
    
    # Load sample and genome trinucleotide background counts, and calculate ratio
    results.trint_counts_and_ratio2genome[[sample_name]] <- read.delim(paste0(dir,"/results.trint_counts_and_ratio2genome.tsv"),header=FALSE,skip=1) %>%
      select(-V3) %>%
      dplyr::rename(tri=V1,sample_tri_bg=V2) %>%
      left_join(genome_trinuc_counts,by="tri") %>%
      mutate(ratio2genome=(sample_tri_bg/sum(sample_tri_bg))/(genome_tri_bg/sum(genome_tri_bg)) )
    
    ratio2genome <- results.trint_counts_and_ratio2genome[[sample_name]] %>% select(tri,ratio2genome)
    
    #Calculate number of observed and corrected SNPs in each trinucleotide context, both for all mutations and after collapsing to unique mutations
    vcf_snp.fix.all <- vcf_snp.fix[[sample_name]] %>% select (CHROM,POS,REF,ALT,INFO) %>%
      mutate(INFO = str_replace(INFO,".*TRI=",""),
      			 INFO = str_replace(INFO,";.*",""),
      			 tri = str_c(str_sub(INFO,1,4),str_sub(INFO,1,1),str_sub(INFO,5,5),str_sub(INFO,3,3))
      ) %>%
    	select(-INFO)
    
    vcf_snp.fix.unique <- vcf_snp.fix.all %>% distinct
    
    vcf_snp.fix.all <- vcf_snp.fix.all %>% select(tri) %>% table %>% as.data.frame %>% rename(trint_subst_observed = Freq)
    vcf_snp.fix.unique <- vcf_snp.fix.unique %>% select(tri) %>% table %>% as.data.frame %>% rename(trint_subst_unique_observed = Freq)
    
    results.trint_subs_obs_corrected[[sample_name]] <- left_join(data.frame(tri = trint_subs_labels),vcf_snp.fix.all,by="tri") %>%
    	left_join(vcf_snp.fix.unique,by="tri") %>%
    	replace(is.na(.), 0) %>%
      mutate(tri_short=str_sub(tri,1,3)) %>%
      left_join(ratio2genome,by=c("tri_short" = "tri")) %>%
      mutate(trint_subst_corrected = trint_subst_observed / ratio2genome,
      			 trint_subst_unique_corrected = trint_subst_unique_observed / ratio2genome) %>%
      select(-tri_short) %>%
      arrange(match(tri,trint_subs_labels))
    
    results.mut_burden[[sample_name]] <- data.frame(
      muts_observed = sum(results.trint_subs_obs_corrected[[sample_name]]$trint_subst_observed),
      muts_corrected = sum(results.trint_subs_obs_corrected[[sample_name]]$trint_subst_corrected),
      total_observed = sum(results.trint_counts_and_ratio2genome[[sample_name]]$sample_tri_bg),
      total_corrected = sum(results.trint_counts_and_ratio2genome[[sample_name]]$sample_tri_bg)
    ) %>% mutate(
      burden_observed = muts_observed / total_observed,
      burden_corrected = muts_corrected / total_corrected,
      muts_lci_observed = poisson.test(muts_observed)$conf.int[1],
      muts_lci_corrected = muts_lci_observed / muts_observed * muts_corrected,
      muts_uci_observed = poisson.test(muts_observed)$conf.int[2],
      muts_uci_corrected = muts_uci_observed / muts_observed * muts_corrected,
      burden_lci_observed = muts_lci_observed / total_observed,
      burden_lci_corrected = muts_lci_corrected / total_corrected,
      burden_uci_observed = muts_uci_observed / total_observed,
      burden_uci_corrected = muts_uci_corrected / total_corrected
    )
    
    results.SSC_mismatches_purine[[sample_name]] <- read.delim(paste0(dir,"/results.SSC-mismatches-Purine.triprofiles.tsv"), header=FALSE) %>%
      mutate(tri = str_c(str_sub(V1,1,4),str_sub(V1,1,1),str_sub(V1,5,5),str_sub(V1,3,3))) %>%
      dplyr::rename(value=V2) %>%
      select(tri,value)
    
    results.SSC_mismatches_pyrimidine[[sample_name]] <- read.delim(paste0(dir,"/results.SSC-mismatches-Pyrimidine.triprofiles.tsv"), header=FALSE) %>%
      mutate(tri = str_c(str_sub(V1,1,4),str_sub(V1,1,1),str_sub(V1,5,5),str_sub(V1,3,3))) %>%
      dplyr::rename(value=V2) %>%
      select(tri,value)
    
    results.estimated_error_rates[[sample_name]] <- read.delim(paste0(dir,"/results.estimated_error_rates.tsv"), header=FALSE, row.names=1) %>% t %>% as.data.frame %>% remove_rownames
  }
  
  message("Combining sample data into data frames...")
  # Collapse lists to data frames
  results.trint_counts_and_ratio2genome <- bind_rows(results.trint_counts_and_ratio2genome,.id="sample")
  results.trint_subs_obs_corrected <- bind_rows(results.trint_subs_obs_corrected,.id="sample")
  results.mut_burden <- bind_rows(results.mut_burden,.id="sample")
  results.SSC_mismatches_purine <- bind_rows(results.SSC_mismatches_purine,.id="sample")
  results.SSC_mismatches_pyrimidine <- bind_rows(results.SSC_mismatches_pyrimidine,.id="sample")
  results.estimated_error_rates <- bind_rows(results.estimated_error_rates,.id="sample")
  
  #Create sigfit format data for observed unique mutation counts and trinucleotide background counts, with samples in rows and trinucleotide contexts in columns
  # Note: using unique mutation counts, since that is a more faithful representation of the mutational process.
  results.trint_subst_obs.sigfit <- results.trint_subs_obs_corrected %>%
    select(sample,tri,trint_subst_unique_observed) %>%
    pivot_wider(names_from=tri,values_from=trint_subst_unique_observed) %>%
    column_to_rownames("sample")
  
  results.trint_counts.sigfit <- results.trint_counts_and_ratio2genome %>%
    select(sample,tri,sample_tri_bg) %>%
    pivot_wider(names_from=tri,values_from=sample_tri_bg) %>%
    column_to_rownames("sample")
  results.trint_counts.sigfit <- results.trint_counts.sigfit[,genome_freqs_labels]
  colnames(results.trint_counts.sigfit) <- genome_freqs_labels
  
  # Create a list or data structure to store the results
  results <- list(
    sample_id = names(vcf_snp.fix),
    vcf_snp.fix = vcf_snp.fix,
    vcf_indel.fix = vcf_indel.fix,
    vcf_indel.gt = vcf_indel.gt,
    trinuc_bg_ratio <- results.trint_counts_and_ratio2genome,
    trinuc_bg.sigfit <- results.trint_counts.sigfit,
    observed_corrected_trinuc_counts <- results.trint_subs_obs_corrected,
    observed_trinuc_counts.sigfit <- results.trint_subst_obs.sigfit,
    mutation_burden <- results.mut_burden,
    purine_trinuc_mismatches <- results.SSC_mismatches_purine,
    pyrimidine_trinuc_mismatches <- results.SSC_mismatches_pyrimidine,
    estimated_error_rates <- results.estimated_error_rates
  )
  
  message("DONE")
  
  return(results)
}