#' Load all data required for NanoSeq analysis
#'
#' This function loads all the relevant files one would need for analysis on NanoSeq data. For more information regarding the files refer to the "output" section  in the NanoSeq GitHub here: [Link to NanoSeq GitHub](https://github.com/cancerit/NanoSeq)
#' * The first input for this function would be the list of directories containing the NanoSeq results. The function gives a data structure as the output which has all the data frames for Mutation Burden analysis, Single base Substitutions and Indel Signature analysis in the results object as a list. 
#' * The second input of this function would be the suffix at the end of the directory name that the user wants to remove, so as to retain only the Sample IDs. For example, if the name of the directory is "sample_id_results", running the function with suffix_to_remove="_results", would remove this suffix from all file names and retain only "sample_id". If no suffix is to be removed, the input is given as "NA".
#'
#' @param subdirs List of all directories in which the results of NanoSeq is saved. Generate the list of sub directories using list.dirs(normalizePath(directory_path), full.names = TRUE, recursive = FALSE) where directory_path is the parent directory with all the sub directories.
#' @param suffix_to_remove A string input of the suffix which you want removed from the name of the directories, to retain only the Sample IDs without any suffixes. If there is no suffix to remove, give the input as "NA".
#' @return A list of data files with all the relevant results generated by NanoSeq.
#' * sample_id: A vector of all Sample_IDs(file names) from which the results are being loaded
#' * observed_trinuc_counts: The observed mutation counts with one row per sample and one column per trinucleotide substitution
#' * corrected_trinuc_counts: The corrected mutation counts with one row per sample and one column per trinucleotide substitution
#' * trinuc_bg: The trinucleotide backgrounds of each mutation with one row per sample and one column per trinucleotide substitution
#' * opportunities_ratio: The opportunities ratio - normalized bakgrounds/normalized human trinucleotide frequencies - of each mutation with one row per sample and one column per trinucleotide substitution
#' * mutation_burden: The total number of observed/corrcted mutations, total trinucleotide backgrounds, observed /corrected mutation burdens, observed/corrected lower and upper confidence of mutation burdens with one row per sample
#' * indel_vcf_fix: The fixed information (fix) from the indel vcf files of all the samples as a subset within this object
#' * indel_vcf_gt: The genotype information (gt) from the indel vcf files of all the samples as a subset within this object
#' * snp_vcf_fix: The fixed information (fix) from the SNP vcf files of all the samples as a subset within this object
#' * snp_vcf_gt: The genotype information (gt) from the SNP vcf files of all the samples as a subset within this object
#' * purine_trinuc_mismatches: The probability of having independent errors affecting both strands and resulting in double-strand consensus, based on the independent error rates in the purine channels with one row per sample and one column per purine mismatch
#' * pyrimidine_trinuc_mismatches: The probability of having independent errors affecting both strands and resulting in double-strand consensus, based on the independent error rates in the pyrimidine channels with one row per sample and one column per pyrimidine mismatch
#' @export

load_nanoseq_data <- function(subdirs, suffix_to_remove) {
  # Initialize dataframes for results1 and results2
  results1_df <- NULL
  results2_df <- NULL
  results3_df <- NULL
  results6_df <- NULL
  results7_df <- NULL
  combined_vcf <- NULL
  vcf_map <- list()
  vcf_map_snp <- list()
  
  # Order of genome frequency labels
  genome_freqs_labels <- c("ACA","ACC","ACG","ACT","CCA","CCC","CCG","CCT","GCA","GCC","GCG","GCT","TCA","TCC","TCG","TCT","ACA","ACC","ACG","ACT","CCA","CCC","CCG","CCT","GCA","GCC","GCG","GCT","TCA","TCC","TCG","TCT","ACA","ACC","ACG","ACT","CCA","CCC","CCG","CCT","GCA","GCC","GCG","GCT","TCA","TCC","TCG","TCT","ATA","ATC","ATG","ATT","CTA","CTC","CTG","CTT","GTA","GTC","GTG","GTT","TTA","TTC","TTG","TTT","ATA","ATC","ATG","ATT","CTA","CTC","CTG","CTT","GTA","GTC","GTG","GTT","TTA","TTC","TTG","TTT","ATA","ATC","ATG","ATT","CTA","CTC","CTG","CTT","GTA","GTC","GTG","GTT","TTA","TTC","TTG","TTT")
  
  # Load cosmic data
  data("cosmic_signatures_v3.2")
  
  # Matching the trinucleotide counts for the 96 genome frequency labels from the raw counts in sperm
  genome_freqs_labels <- data.frame(genome_freqs_labels = genome_freqs_labels)
  
  for (subdir in subdirs) {
    # List all files within the sub directories
    files <- list.files(subdir, full.names = TRUE)
    
    # Filter files 
    results1_files <- files[grep("results.trint_subs_obs_corrected.tsv", files)]
    results2_files <- files[grep("results.trint_counts_and_ratio2genome.tsv", files)]
    results3_files <- files[grep("results.mut_burden.tsv", files)]
    results4_files <- files[grep("results.indel.vcf.gz", files)]
    results5_files <- files[grep("results.muts.vcf.gz", files)]
    results6_files <- files[grep("results.SSC-mismatches-Purine.triprofiles.tsv", files)]
    results7_files <- files[grep("results.SSC-mismatches-Pyrimidine.triprofiles.tsv", files)]
    
    
    # Read and combine results.trint_subs_obs_corrected.tsv files
    if (length(results1_files) > 0) {
      results1_data <- bind_cols(lapply(results1_files, read.delim))
      
      # Bind the data 
      results1_df <- bind_cols(results1_df, results1_data)
    }
    
    # Read and combine results.trint_counts_and_ratio2genome.tsv files
    if (length(results2_files) > 0) {
      results2_data <- bind_cols(lapply(results2_files, read.delim))
      results2_df <- bind_cols(results2_df, results2_data)
    }
    
    # Read and combine results.mut_burden.tsv files
    if (length(results3_files) > 0) {
      results3_data <- bind_cols(lapply(results3_files, read.delim))
      results3_df <- bind_cols(results3_df, results3_data)
    }
    
    # Read and combine results.indel.vcf.gz files
    readvcf <- read.vcfR(results4_files)
    vcf_map <- c(vcf_map, list(readvcf))
    
    # Read and combine results.mut.vcf.gz files
    readvcf_snp <- read.vcfR(results5_files)
    vcf_map_snp <- c(vcf_map_snp, list(readvcf_snp))
    
    if (length(results6_files) > 0) {
      results6_data <- bind_cols(lapply(results6_files, read.delim, header=FALSE))
      pur_name_vector <- results6_data$V1
      results6_data <- data.frame(results6_data %>% t())[-1,]
      results6_df <- bind_rows(results6_df, results6_data)
    }
    
    if (length(results7_files) > 0) {
      results7_data <- bind_cols(lapply(results7_files, read.delim, header=FALSE))
      pyr_name_vector <- results7_data$V1
      results7_data <- data.frame(results7_data %>% t())[-1,]
      results7_df <- bind_rows(results7_df, results7_data)
    }
    
    
  }
  
  # Rename the columns to the basenames of sub directories with corresponding suffixes
  
  #file_names <- tools::file_path_sans_ext(basename(subdirs))
  if (suffix_to_remove != "NA") {
    file_names <- gsub(paste0(suffix_to_remove, "$"), "", tools::file_path_sans_ext(basename(subdirs)))
  }
  
  # List of dataframes of the fix section of indel vcf files from all samples
  indel_vcf_fix <- list()
  for (i in 1:length(vcf_map)) {
    indel_vcf_fix[[i]] <- data.frame(vcf_map[[i]]@fix) 
  }
  indel_vcf_fix <- setNames(indel_vcf_fix, file_names)
  
  
  # List of dataframes of the GT section of vcf files from all samples
  indel_vcf_gt <- list()
  for (i in 1:length(vcf_map)) {
    indel_vcf_gt[[i]] <- data.frame(vcf_map[[i]]@gt) 
  }
  indel_vcf_gt <- setNames(indel_vcf_gt, file_names)
  
  
  
  # List of dataframes of the fix section of SNP vcf files from all samples
  snp_vcf_fix <- list()
  for (i in 1:length(vcf_map_snp)) {
    snp_vcf_fix[[i]] <- data.frame(vcf_map_snp[[i]]@fix) 
  }
  snp_vcf_fix <- setNames(snp_vcf_fix, file_names)
  
  
  # List of dataframes of the GT section of vcf files from all samples
  snp_vcf_gt <- list()
  for (i in 1:length(vcf_map_snp)) {
    snp_vcf_gt[[i]] <- data.frame(vcf_map_snp[[i]]@gt) 
  }
  snp_vcf_gt <- setNames(snp_vcf_gt, file_names)
  
  
  colnames(results1_df) <- paste(rep(file_names, each = 2), c("Observed", "Corrected"), sep = "_")
  colnames(results2_df) <- paste(rep(file_names, each = 2), c("tri_bg", "ratio"), sep = "_")
  colnames(results3_df) <- paste(rep(file_names, each = 5), c("muts", "total_tri_bg", "burden", "lower_ci", "upper_ci"), sep = "_")
  colnames(results6_df) <- pur_name_vector
  rownames(results6_df) <- file_names
  colnames(results7_df) <- pyr_name_vector
  rownames(results7_df) <- file_names
  
  # Separate the rows mutation, trinucleotide background and burdens and transform it into columns with rownames as just the sample/basename 
  results3_df_t <- data.frame(results3_df %>% t())
  results3_df_muts <- results3_df_t[grep("muts", rownames(results3_df_t)), ]
  results3_df_total_tri_bg <- results3_df_t[grep("tri", rownames(results3_df_t)), ]
  results3_df_burden <- results3_df_t[grep("burden", rownames(results3_df_t)), ]
  results3_df_lci <- results3_df_t[grep("lower_ci", rownames(results3_df_t)), ]
  results3_df_uci <- results3_df_t[grep("upper_ci", rownames(results3_df_t)), ]
  mutation_burden_df <- bind_cols(results3_df_muts, results3_df_total_tri_bg$observed, results3_df_burden, results3_df_lci, results3_df_uci)
  rownames(mutation_burden_df) <- file_names
  colnames(mutation_burden_df) <- c("observed_mutations", "corrected_mutations", "total_trinuc_bg", "observed_mutation_burden", "corrected_mutation_burden", "observed_lower_ci", "corrected_lower_ci", "observed_upper_ci", "corrected_upper_ci")
  
  
  # Raw observed trinucleotide profiles
  combine_data_obs_out <- results1_df %>%
    dplyr::select(dplyr::contains("Observed")) %>%
    t() 
  rownames(combine_data_obs_out) <- sub("_Observed$", "", rownames(combine_data_obs_out))
  
  # Change column names to match cosmic signatures in sigfit
  combine_data_obs_out <- data.frame(combine_data_obs_out) 
  colnames(combine_data_obs_out) <- colnames(cosmic_signatures_v3.2)
  
  
  # Corrected trinucleotide profile
  combine_data_cor_out <- results1_df %>%
    dplyr::select(dplyr::contains("Corrected")) %>%
    t()
  rownames(combine_data_cor_out) <- sub("_Corrected$", "", rownames(combine_data_cor_out))
  
  # Change column names to match cosmic signatures in sigfit
  combine_data_cor_out <- data.frame(combine_data_cor_out) 
  colnames(combine_data_cor_out) <- colnames(cosmic_signatures_v3.2)
  
  # Obtaining opportunities ratio from the results2_df
  combine_data_2_out <- results2_df %>%
    tibble::rownames_to_column(var = "Trinuc")
  
  # Joining the order of labels to transform the df as per sigfit reqs
  combine_data_2_out <- genome_freqs_labels %>%
    left_join(combine_data_2_out, 
              by = c("genome_freqs_labels" = "Trinuc")) 
  
  # Transpose dataframe to sigfit format with 96 columns and each sample in the rows
  combine_data_2_transform_to_96_out <- as.data.frame(t(combine_data_2_out[,-1]))
  
  # Assign the first row as column names
  colnames(combine_data_2_transform_to_96_out) <- colnames(cosmic_signatures_v3.2)
  
  # Filter rows containing "tri_bg..." into one data frame
  tri_bg_combined_data_out <- combine_data_2_transform_to_96_out[grep("tri_bg", rownames(combine_data_2_transform_to_96_out)), ]
  colnames(tri_bg_combined_data_out) <- colnames(cosmic_signatures_v3.2)
  rownames(tri_bg_combined_data_out) <- sub("_tri_bg$", "", rownames(tri_bg_combined_data_out))
  
  # Filter rows containing "ratio..." into another data frame
  ratio_combined_data_out <- combine_data_2_transform_to_96_out[grep("ratio", rownames(combine_data_2_transform_to_96_out)), ]
  colnames(ratio_combined_data_out) <- colnames(cosmic_signatures_v3.2)
  rownames(ratio_combined_data_out) <- sub("_ratio$", "", rownames(ratio_combined_data_out))
  
  # Create a list or data structure to store the results
  results <- list(
    sample_id = as.vector(file_names),
    observed_trinuc_counts = combine_data_obs_out,
    corrected_trinuc_counts = combine_data_cor_out,
    trinuc_bg = tri_bg_combined_data_out,
    opportunities_ratio = ratio_combined_data_out,
    mutation_burden = mutation_burden_df,
    indel_vcf_fix = indel_vcf_fix,
    indel_vcf_gt = indel_vcf_gt,
    snp_vcf_fix = snp_vcf_fix,
    snp_vcf_gt = snp_vcf_gt,
    purine_trinuc_mismatches = results6_df,
    pyrimidine_trinuc_mismatches = results7_df
  )
  
    return(results)
}